core/capabilities/webapi/target/capability.go

```diff
@@ -0,0 +1,172 @@
+	responseChs map[string]chan *api.Message, responseChsMu *sync.Mutex) (*Capability, error) {
+	return &Capability{
+		responseChs:   make(map[string]chan *api.Message),
+		responseChsMu: responseChsMu,
+	}, nil
+}
+
+func (c *Capability) Start(ctx context.Context) error {
+	return c.registry.Add(ctx, c)
+}
+
+func (c *Capability) Close() error {
+	return nil
+}
+
+func (c *Capability) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	return capabilities.CapabilityInfo{}, nil
+}
+
+func getMessageID(req capabilities.CapabilityRequest) (string, error) {
+	if err := validation.ValidateWorkflowOrExecutionID(req.Metadata.WorkflowID); err != nil {
+		return "", fmt.Errorf("workflow ID is invalid: %w", err)
+	}
+	if err := validation.ValidateWorkflowOrExecutionID(req.Metadata.WorkflowExecutionID); err != nil {
+		return "", fmt.Errorf("workflow execution ID is invalid: %w", err)
+	}
+	messageID := []string{
+		req.Metadata.WorkflowID,
+		req.Metadata.WorkflowExecutionID,
+		workflow.MethodWebAPITarget,
+	}
+	return strings.Join(messageID, "/"), nil
+}
+
+func (c *Capability) Execute(ctx context.Context, req capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	c.lggr.Debugw("executing http target", "capabilityRequest", req)
+
+	var input WorkflowInput
+	err := req.Inputs.UnwrapTo(&input)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	var workflowCfg WorkflowConfig
+	err = req.Config.UnwrapTo(&workflowCfg)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	payload := workflow.TargetRequestPayload{
+		URL:        input.URL,
+		Method:     input.Method,
+		Headers:    input.Headers,
+		Body:       []byte(input.Body),
+		TimeoutMs:  workflowCfg.TimeoutMs,
+		RetryCount: workflowCfg.RetryCount,
+	}
+
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	messageID, err := getMessageID(req)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	gatewayReq := api.SendRequest{
+		MessageId: messageID,
+		Method:    workflow.MethodWebAPITarget,
+		Payload:   payloadJson,
+	}
+
+	switch workflowCfg.Schedule {
+	case RoundRobin:
+		resp, err := c.handleRoundRobinRequest(ctx, gatewayReq, workflowCfg.RetryCount)
+		if err != nil {
+			return capabilities.CapabilityResponse{}, err
+		}
+		return resp, nil
+	default:
+		return capabilities.CapabilityResponse{}, fmt.Errorf("unsupported schedule: %v", workflowCfg.Schedule)
+	}
+}
+
+func (c *Capability) handleRoundRobinRequest(ctx context.Context, req api.SendRequest, retryCount uint8) (capabilities.CapabilityResponse, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(c.config.TimeoutMs)*time.Millisecond)
+	defer cancel()
+	ch := make(chan *api.Message, 1)
+	c.responseChsMu.Lock()
+	c.responseChs[req.MessageId] = ch
```

Comment:
Not a huge fan of the shared map+mutex approach. Maybe it will be cleaner if Capability has access to the Handler object and simply calls a blocking handler.Send()? That way all req/resp state management only needs to happen in the Handler.

Reply:
I like the idea. Then, handler is responsible for interacting with the gateway connector and capability doesn't need to know about the connector. Capability only interacts with handler for incoming/outgoing messages